cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapReduceTransformExecutorFactory.java

```diff
@@ -101,11 +107,94 @@ protected TrackedTransform getTransformation(String pluginType, String stageName
+   * in mapreduce.
+   *
+   * @param <JOIN_KEY> type of join key
+   * @param <JOIN_VALUE> type of the join value
+   * @param <OUT> type of the output of joiner
+   * @param <OUT_KEY> type of the map output
+   */
+  private static class MapperJoinerTransformation<JOIN_KEY, JOIN_VALUE, OUT, OUT_KEY extends Writable>
+    implements Transformation<JOIN_VALUE, KeyValue<OUT_KEY, MapTaggedOutputWritable<JOIN_VALUE>>> {
+    private final Joiner<JOIN_KEY, JOIN_VALUE, OUT> joiner;
+    private final WritableConversion<JOIN_KEY, OUT_KEY> keyConversion;
+
+    MapperJoinerTransformation(Joiner<JOIN_KEY, JOIN_VALUE, OUT> joiner, String joinKeyClassName) {
+      this.joiner = joiner;
+      WritableConversion<JOIN_KEY, OUT_KEY> keyConversion = WritableConversions.getConversion(joinKeyClassName);
+      this.keyConversion = keyConversion == null ? new CastConversion<JOIN_KEY, OUT_KEY>() : keyConversion;
+    }
+
+    @Override
+    public void transform(JOIN_VALUE input, Emitter<KeyValue<OUT_KEY, MapTaggedOutputWritable<JOIN_VALUE>>> emitter)
+      throws Exception {
+      String stageName;
+      // TODO Use input to get stageName
+      if (emitter instanceof TransformDetail) {
+        stageName = ((TransformDetail) emitter).getPrevStage();
+      } else {
+        stageName = ((TransformDetail) ((TrackedEmitter) emitter).getEmitter()).getPrevStage();
+      }
+      JOIN_KEY key = joiner.joinOn(stageName, input);
+      MapTaggedOutputWritable<JOIN_VALUE> output = new MapTaggedOutputWritable(stageName, input);
+      emitter.emit(new KeyValue<>(keyConversion.toWritable(key), output));
+    }
+  }
+
+  /**
+   * A Transformation that uses an join's merge method based on type of the join.
+   *
+   * @param <JOIN_KEY> type of join key
+   * @param <JOIN_VALUE> type of the join value
+   * @param <OUT> type of the output of joiner
+   * @param <REDUCE_KEY> type of the reduce key=
+   */
+  private static class ReducerJoinerTransformation<JOIN_KEY, JOIN_VALUE, OUT, REDUCE_KEY extends WritableComparable>
+    implements Transformation<KeyValue<REDUCE_KEY, Iterator<MapTaggedOutputWritable<JOIN_VALUE>>>, OUT> {
+    private final Joiner<JOIN_KEY, JOIN_VALUE, OUT> joiner;
+    private final WritableConversion<JOIN_KEY, REDUCE_KEY> keyConversion;
+
+    ReducerJoinerTransformation(Joiner<JOIN_KEY, JOIN_VALUE, OUT> joiner, String joinKeyClassName) {
+      this.joiner = joiner;
+      WritableConversion<JOIN_KEY, REDUCE_KEY> keyConversion = WritableConversions.getConversion(joinKeyClassName);
+      this.keyConversion = keyConversion == null ? new CastConversion<JOIN_KEY, REDUCE_KEY>() : keyConversion;
+    }
+
+    @Override
+    public void transform(KeyValue<REDUCE_KEY, Iterator<MapTaggedOutputWritable<JOIN_VALUE>>> input,
+                          Emitter<OUT> emitter) throws Exception {
+      JOIN_KEY joinKey = keyConversion.fromWritable(input.getKey());
+      JoinConfig joinConfig = joiner.getJoinConfig();
```

Comment:
this part is confusing to me. Shouldn't we just be calling the Joiner's merge method instead checking the join type and switching on it? 

I thought the purpose of the merge method was to do inner, left, outer types of joins or is that incorrect?